### PR TITLE
Add Deny List Feature & Add Reasons to `limit` Response

### PR DESCRIPTION
**Pull request by CahidArda**
## Deny List

In redis, we keep sets of denied values. One set each for identifiers, ips, countries and user agents. Finally, there is a set of all blocked values. Upstash console is responsible for managing these sets. Ratelimit sdk is only interested in the set of all blocked values stored in key `<prefix>:denyList:all`.

Enabling deny list in the sdk:

```tsx
const ratelimit = new Ratelimit({
  redis: Redis.fromEnv(),
  limiter: Ratelimit.slidingWindow(10, "10 s"),
  enableProtection: true // set to true
});
```

Passing ip, user agent or country in `limit`:

```tsx
ratelimit.limit(
  "identifier",
  {
    ip: "12.32.12.32",
    userAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X x.y; rv:42.0) Gecko/20100101 Firefox/42.0",
    country: "Germany"
  }
)
```

all three fields are optional. One can pass none/some/all.

Inside limit, we check if any of the three fields or the identifier is in the deny list. If so, we return `{success: false, reason: "denyList", ...}`

### Deny List Cache

When a value is found in the deny list at redis, we store the value for a minute in a local cache. If that value is passed to the `limit` method in the next minute, we block the request with `reason: "denyList"` without making any requests to Redis.

## `reason` field in `limit` response

We add a new field to the response of `limit`, which can be either "timeout", "cacheBlock" or "denyList". In the case of "denyList", there is an additional field `_deniedValue ` which denotes the value in deny list.

### Deny List Analytics

When a request is denied, we submit increment `{identifier: <deniedValue>, success:"denied"}` by one.

Previously:
- the identifier could only be the identifier value passed to `limit`. Now it can be one of the 4 parameters.
- success could only be true or false. Now we allow "denied" to count the denied responses.

## Todo:

- [x] tests
- [x] documentation
- [ ] bump core analytics after https://github.com/upstash/core-analytics/pull/37

### CODE REVIEW

File: src/analytics.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export type Geo = {
 export type Event = Geo & {
   identifier: string;
   time: number;
-  success: boolean;
+  success: boolean | string;
```
Comment (ogzhanolguncu) [inline src/analytics.ts:13]: Comment or Branded type highly appreciated.

File: src/denyList.ts
Comment (ogzhanolguncu) [inline src/denyList.ts:1]: let's keep the file name convention same: `deny-list.ts`

File: src/denyList.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * Checks items in members list and returns true if any
+ * of them are in denyListCache.
+ * 
+ * @param members list of values to check against the cache
+ * @returns a member from the cache. If there is none, returns undefined
+ */
+export const checkDenyListCache = (members: string[]): string | undefined => {
+  return members.filter(
```
Comment (ogzhanolguncu) [inline src/denyList.ts:16]: `some()` would be a better choice here. If our goal is to 'get any of them'

File: src/denyList.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  prefix: string,
+  members: string[]
+) => {
+  const deniedMembers = await redis.smismember(
+    [prefix, "denyList", "all"].join(":"),
+    members
+  );
+
+  let deniedMember: string | undefined = undefined;
+  deniedMembers.map((memberDenied, index) => {
```
Comment (ogzhanolguncu) [inline src/denyList.ts:56]: Maybe we can try a simple `for-of` instead of map, since we are not actually mapping anything?

File: src/types.ts [hunk 4]
```diff
@@ -98,4 +109,8 @@ export interface Redis {
   scriptLoad: (
     ...args: [script: string]
   ) => Promise<string>;
+
+  smismember: (
+    key: string, members: string[]
+  ) => Promise<(0 | 1)[]>;
```
Comment (ogzhanolguncu) [inline src/types.ts:115]: Maybe we can put another branded type here.

File: src/types.ts [hunk 5]
```diff
@@ -64,6 +68,13 @@ export type RatelimitResponse = {
    * ```
    */
   pending: Promise<unknown>;
+
+  reason?: ResponseType;
+
+  /**
+   * The value which was in the deny list if reason: "denied"
+   */
+  _deniedValue?: string
```
Comment (ogzhanolguncu) [inline src/types.ts:77]: Is there any reason why this is underscored? I believe it is because this is an internal value, but I'd omit the underscore to maintain the convention.

File: src/denyList.ts (hunk 2 above)
Comment (CahidArda) [reply src/denyList.ts:16]: did it using find()

File: src/types.ts (hunk 4 above)
Comment (CahidArda) [reply src/types.ts:115]: added IsMember

File: src/types.ts (hunk 5 above)
Comment (CahidArda) [reply src/types.ts:77]: I meant that it's for internal use, but it's okay to remove the underscore

File: src/analytics.ts (hunk 1 above)
Comment (CahidArda) [reply src/analytics.ts:13]: defined type EventSuccess. used "denied" instead of string. I initially thought using string would make it more flexible, but writing it in a restrictive way and extending later if needed makes more sense.

File: src/analytics.ts [hunk 6]
```diff
@@ -7,10 +7,18 @@ export type Geo = {
   region?: string;
   ip?: string;
 };
+
+/**
+ * denotes the success field in the analytics submission.
+ * Set to true when ratelimit check passes. False when request is ratelimited.
+ * Set to "denied" when some request value is in deny list.
+ */
+export type EventSuccess = boolean | "denied"
```
Comment (ogzhanolguncu) [inline src/analytics.ts:16]: You know what `"pass" | "ratelimited" | "denied"` would be fantastic IMO. Definitely more expressive.

File: src/ratelimit.ts [hunk 7]
```diff
@@ -148,37 +165,64 @@ export abstract class Ratelimit<TContext extends Context> {
...
+      ip?: string,
+      userAgent?: string,
+      country?: string
+    },
   ): Promise<RatelimitResponse> => {
     const key = [this.prefix, identifier].join(":");
     let timeoutId: any = null;
+
+    const members = [identifier, req?.ip, req?.userAgent, req?.country];
+    const definedMembers = members.filter(item => item !== undefined) as string[];
```
Comment (ogzhanolguncu) [inline src/ratelimit.ts:180]: Let's not cast that like `as string[]` this `as` might mask some error in the future. Let's properly type this.

File: src/ratelimit.ts [hunk 8]
```diff
@@ -148,37 +165,64 @@ export abstract class Ratelimit<TContext extends Context> {
...
+                    pending: Promise.resolve(),
+                    reason: "timeout"
+                  },
+                  undefined
+                ]);
+              }, this.timeout);
+            }),
+          );
+        }
+        res = resolveResponses(...(await Promise.race(arr)));
```
Comment (ogzhanolguncu) [inline src/ratelimit.ts:214]: I think its time to break this `limit()` function into smaller pieces. Its starting to get really hard to follow. Also, having some explanatory comments would be fantastic, wdyt? We don't have do this right away.

File: src/denyList.ts (hunk 2 above)
Comment (ogzhanolguncu) [reply src/denyList.ts]: Ah, sorry I thought you were just checking if its exists or not. Good call.

File: src/analytics.ts (hunk 6 above)
Comment (CahidArda) [reply src/analytics.ts:16]: this would also fix the [`1 | null | denied` in core analytics](https://github.com/upstash/core-analytics/pull/37#discussion_r1620341827), but it would be backwards incompatible, users wouldn't be able to view past analytics at all after we update the dashboard.

I think we can do it if you are okay with backwards compatibility

File: src/ratelimit.ts (hunk 8 above)
Comment (CahidArda) [reply src/ratelimit.ts]: I also thought that it was getting out of control. Did some refactoring and I think now it looks way cleaner.

File: src/ratelimit.ts (hunk 7 above)
Comment (CahidArda) [reply src/ratelimit.ts]: updated as `members.filter((item): item is string => Boolean(item))`

File: src/analytics.ts (hunk 6 above)
Comment (ogzhanolguncu) [reply src/analytics.ts:16]: Agreed. Let's not forget to create a ticket for this.

File: src/ratelimit.ts (hunk 8 above)
Comment (ogzhanolguncu) [reply src/ratelimit.ts]: 10x better. I like it.

File: src/analytics.ts (hunk 6 above)
Comment (CahidArda) [reply src/analytics.ts:16]: [ticket](https://linear.app/upstash/issue/DX-966/ratelimit-success-conditions-update)

**Review by ogzhanolguncu — APPROVED** [review]

**Comment by CahidArda** [comment]
I think we can go ahead with this PR even though the tests are not passing because they are flaky. Here are two runs with two different tests failing even though the code didn't change between them: [1](https://github.com/upstash/ratelimit-js/actions/runs/9316584645/job/25653266192), [2](https://github.com/upstash/ratelimit-js/actions/runs/9316584645/job/25652249766).

I looked into [recording network traffic like langchain does](https://github.com/amosjyng/vcr-langchain) but that's only in python. I want to fix the tests but let's do it another time
